src/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateStream.cs

```diff
@@ -0,0 +1,649 @@
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException(SR.NotSupported);
+        }
+
+        public override int Read(byte[] array, int offset, int count)
+        {
+            EnsureDecompressionMode();
+            ValidateParameters(array, offset, count);
+            EnsureNotDisposed();
+
+            int bytesRead;
+            int currentOffset = offset;
+            int remainingCount = count;
+
+            while (true)
+            {
+                bytesRead = _inflater.Inflate(array, currentOffset, remainingCount);
+                currentOffset += bytesRead;
+                remainingCount -= bytesRead;
+
+                if (remainingCount == 0)
+                {
+                    break;
+                }
+
+                if (_inflater.Finished())
+                {
+                    // if we finished decompressing, we can't have anything left in the outputwindow.
+                    Debug.Assert(_inflater.AvailableOutput == 0, "We should have copied all stuff out!");
+                    break;
+                }
+
+                int bytes = _stream.Read(_buffer, 0, _buffer.Length);
+                if (bytes <= 0)
+                {
+                    break;
+                }
+                else if (bytes > _buffer.Length)
+                {
+                    // The stream is either malicious or poorly implemented and returned a number of
+                    // bytes larger than the buffer supplied to it.
+                    throw new InvalidDataException(SR.GenericInvalidData);
+                }
+
+                _inflater.SetInput(_buffer, 0, bytes);
+            }
+
+            return count - remainingCount;
+        }
+
+        private void ValidateParameters(byte[] array, int offset, int count)
+        {
+            if (array == null)
+                throw new ArgumentNullException("array");
+
+            if (offset < 0)
+                throw new ArgumentOutOfRangeException("offset");
+
+            if (count < 0)
+                throw new ArgumentOutOfRangeException("count");
+
+            if (array.Length - offset < count)
+                throw new ArgumentException(SR.InvalidArgumentOffsetCount);
+        }
+
+        private void EnsureNotDisposed()
+        {
+            if (_stream == null)
+                throw new ObjectDisposedException(null, SR.ObjectDisposed_StreamClosed);
+        }
+
+        private void EnsureDecompressionMode()
+        {
+            if (_mode != CompressionMode.Decompress)
+                throw new InvalidOperationException(SR.CannotReadFromDeflateStream);
+        }
+
+        private void EnsureCompressionMode()
+        {
+            if (_mode != CompressionMode.Compress)
+                throw new InvalidOperationException(SR.CannotWriteToDeflateStream);
+        }
+
+        public override Task<int> ReadAsync(Byte[] array, int offset, int count, CancellationToken cancellationToken)
```

Comment:
@joperezr had done some work to add Begin/End overrides to System.IO.Compression.  Does any of the code added back need those overrides added as well?